src/analyses/apron/apronAnalysis.apron.ml

```diff
@@ -300,14 +352,40 @@ struct
     else
       unify_st
 
+
+  let invalidate_one ask ctx st lv =
+    assign_to_global_wrapper ask ctx.global ctx.sideg st lv (fun st v ->
+        let apr' = AD.forget_vars st.apr [V.local v] in
+        assert_type_bounds apr' v (* re-establish type bounds after forget *)
+      )
+
+
+  (* Give the set of reachables from argument. *)
+  let reachables (ask: Queries.ask) es =
+    let reachable e st =
+      match st with
+      | None -> None
+      | Some st ->
+        let vs = ask.f (Queries.ReachableFrom e) in
+        if Queries.LS.is_top vs then
+          None
+        else
+          Some (Queries.LS.join vs st)
+    in
+    List.fold_right reachable es (Some (Queries.LS.empty ()))
+
+
+  let forget_reachable ctx st es =
+    let ask = Analyses.ask_of_ctx ctx in
+    match reachables ask es with
+    | None -> D.top ()
```

Comment:
On knot from bench it seems like this `D.top ()` might have caused a crash since `DHetero.top` is defined to fail. I think it might have to be `D.top_env` (using the previous environment), because we still need to have the environment but forget everything about it.

Reply:
Actually one can't just do that because something needs to happen with the `priv` component as well. Maybe we can just use `invalidate_one` on all the variables in the environment then? That does a self-write which privatization is aware of.

Reply:
The question is do we have a convenient way to get all the CIL(!) variables in the environment here? `invalidate_one` expects a CIL Lval...

Reply:
I committed a fix that at least avoids the crash. The primed Arg variables and the Return variable aren't actually reachable so shouldn't be invalidated, therefore `to_cil_varinfo` should suffice.